include/kernels/color_spinor_pack.cuh

```diff
@@ -1,206 +1,243 @@
-    constexpr int thread_width_x = ( (max_block_size / color_spin_threads + bank_width-1) / bank_width) * bank_width;
-    __shared__ Float v[ (Ns/Ms) * (Nc/Mc) * thread_width_x];
-    const auto &rhs = arg.field;
-    if (active) {
-#pragma unroll
-      for (int spin_local=0; spin_local<Ms; spin_local++) {
-	int s = spin_block + spin_local;
+  template <> struct site_max<true> {
+    template <typename Arg> __device__ inline auto operator()(typename Arg::real thread_max, Arg &)
+    {
+      using real = typename Arg::real;
+      constexpr int Ms = spins_per_thread<true, Arg::nSpin>();
+      constexpr int Mc = colors_per_thread<true, Arg::nColor>();
+      constexpr int color_spin_threads = (Arg::nSpin/Ms) * (Arg::nColor/Mc);
+      SharedMemoryCache<real, color_spin_threads, 2, true> cache; // 2 comes from parity
+      cache.save(thread_max);
+      cache.sync();
+      real this_site_max = static_cast<real>(0);
 #pragma unroll
-	for (int color_local=0; color_local<Mc; color_local++) {
-	  int c = color_block + color_local;
-	  complex<Float> z = rhs(spinor_parity, x_cb, s, c);
-	  thread_max = thread_max > fabs(z.real()) ? thread_max : fabs(z.real());
-	  thread_max = thread_max > fabs(z.imag()) ? thread_max : fabs(z.imag());
-	}
+      for (int sc = 0; sc < color_spin_threads; sc++) {
+        auto sc_max = cache.load_y(sc);
+        this_site_max = this_site_max > sc_max ? this_site_max : sc_max;
       }
-      v[ ( (spin_block/Ms) * (Nc/Mc) + (color_block/Mc)) * blockDim.x + threadIdx.x ] = thread_max;
+      return this_site_max;
     }
+  };
+
+  template <typename Arg> __device__ __host__ inline std::enable_if_t<!Arg::block_float, typename Arg::real>
+  compute_site_max(const Arg &, int, int, int, int)
+  {
+    return static_cast<typename Arg::real>(1.0); // dummy return for non-block float
+  }
+
+  /**
+     Compute the max element over the spin-color components of a given site.
+  */
+  template <typename Arg> __device__ __host__ inline std::enable_if_t<Arg::block_float, typename Arg::real>
+  compute_site_max(const Arg &arg, int x_cb, int spinor_parity, int spin_block, int color_block)
+  {
+    using real = typename Arg::real;
+    const int Ms = spins_per_thread<Arg::nSpin>();
+    const int Mc = colors_per_thread<Arg::nColor>();
+    real thread_max_r = 0.0;
+    real thread_max_i = 0.0;
 
-    __syncthreads();
-   
-    if (active) {
 #pragma unroll
-      for (int sc=0; sc<(Ns/Ms) * (Nc/Mc); sc++) {
-	site_max = site_max > v[sc*blockDim.x + threadIdx.x] ? site_max : v[sc*blockDim.x + threadIdx.x];
+    for (int spin_local=0; spin_local<Ms; spin_local++) {
+      int s = spin_block + spin_local;
+#pragma unroll
+      for (int color_local=0; color_local<Mc; color_local++) {
+        int c = color_block + color_local;
+        complex<real> z = arg.field(spinor_parity, x_cb, s, c);
+        thread_max_r = std::max(thread_max_r, std::abs(z.real()));
```

Comment:
Should this be `target::dispatch`? (Noting target dispatch gets used for the return)

Reply:
Discussed offline.  This was a misunderstanding.  Marking as resolved.